cmd/mirror-main.go

```diff
@@ -522,16 +516,148 @@ func newMirrorJob(ctx *cli.Context) *mirrorJob {
+				}
+				mj.mirror()
+				wg.Done()
+			}
+		}(i)
+	}
+
+	srcClt, _ := newS3Client(srcURL)
+	dstClt, _ := newS3Client(dstURL)
+
+	// Synchronize buckets using dirDifference function, if preserve-policies
+	// is enabled, copy policies rules.
+	for d := range dirDifference(srcClt, dstClt, srcURL, dstURL) {
+		if d.Diff == differInSecond {
+			// Ignore buckets that only exist in target instance
+			continue
+		}
+
+		// Construct new clients which point to buckets
+		newSrcURL := srcURL + d.firstContent.URL.Path
+		newTgtURL := dstURL + d.firstContent.URL.Path
+
+		newSrcClt, _ := newS3Client(newSrcURL)
+		newDstClt, _ := newS3Client(newTgtURL)
+
+		if d.Diff == differInFirst {
+			// Bucket only exists in the source, create the same bucket in the destination
+			err := newDstClt.MakeBucket(ctx.String("region"))
+			fatalIf(err, "Cannot created bucket in `"+newTgtURL+"`")
+		}
+
+		// Copy policy rules from source to dest if flag is activated
+		if ctx.Bool("preserve-policies") {
+			rules, err := newSrcClt.GetAccessRules()
+			fatalIf(err, "Cannot get policy rules")
+			// Set found rules to target bucket if permitted
+			for _, r := range rules {
+				originalRule, err := newDstClt.GetAccess()
+				fatalIf(err, "Cannot get bucket policy rules from `"+newTgtURL+"`")
+				// Set rule only if it doesn't exist in the target bucket
+				// or force flag is activated
+				if originalRule == "none" || ctx.Bool("force") {
+					err = newDstClt.SetAccess(r)
+					fatalIf(err, "Cannot set new bucket policy rules to `"+newTgtURL+"`")
+				}
+			}
+		}
+
+		// Send the mirror job to any worker
+		wg.Add(1)
+		work <- &mirrorInfo{source: newSrcURL, target: newTgtURL}
+	}
+
+	wg.Wait()
+	close(work)
+
+	return nil
+}
+
+// mirrorSingleBucket
+func mirrorSingleBucket(srcURL, tgtURL string, ctx *cli.Context) error {
+	mj := newMirrorJob(srcURL, tgtURL, 1,
+		ctx.Bool("fake"),
+		ctx.Bool("remove"),
+		ctx.Bool("watch"),
+		ctx.Bool("force"),
+	)
+
+	// Starts additional watcher thread for watching for new events.
+	if mj.isWatch {
+		// Start watching and mirroring.
+		go mj.watchMirror()
+	}
+
+	mj.mirror()
+	return nil
+}
+
 // Main entry point for mirror command.
 func mainMirror(ctx *cli.Context) error {
+
 	// check 'mirror' cli arguments.
 	checkMirrorSyntax(ctx)
 
 	// Additional command specific theme customization.
 	console.SetColor("Mirror", color.New(color.FgGreen, color.Bold))
 
-	mj := newMirrorJob(ctx)
-	mj.mirror()
+	args := ctx.Args()
+
+	srcURL := args[0]
+	tgtURL := args[1]
+
+	srcClt, err := newClient(srcURL)
+	fatalIf(err, "Unable to initialize `"+srcURL+"`")
+
+	dstClt, err := newClient(tgtURL)
+	fatalIf(err, "Unable to initialize `"+tgtURL+"`")
+
+	mirrorAll := false
+
+	// Check if passed arguments could trigger a complete mirror instance
+	if srcClt.GetURL().Type == objectStorage && dstClt.GetURL().Type == objectStorage {
+		// Check if buckets are not passed
+		srcBucket, _ := srcClt.(*s3Client).url2BucketAndObject()
```

Comment:
> We should never have code like this.. unless really really needed. what problem are you trying to solve here?

Here, I needed to check if the user passed any buckets in arguments to know if we are going to mirror all buckets or it is just a simple mirror